src/main/java/com/artipie/metrics/asto/StorageMetrics.java

```diff
@@ -0,0 +1,148 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.artipie.metrics.asto;
+
+import com.artipie.asto.Content;
+import com.artipie.asto.Key;
+import com.artipie.asto.Storage;
+import com.artipie.asto.ext.PublisherAs;
+import com.artipie.metrics.Counter;
+import com.artipie.metrics.Gauge;
+import com.artipie.metrics.Metrics;
+import com.jcabi.log.Logger;
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+/**
+ * Storage metrics.
+ * @since 0.19
+ */
+public final class StorageMetrics implements Metrics {
+
+    /**
+     * Storage for metrics.
+     */
+    private final Storage storage;
+
+    /**
+     * New storage metrics.
+     * @param storage Storage
+     */
+    public StorageMetrics(final Storage storage) {
+        this.storage = storage;
+    }
+
+    @Override
+    public Counter counter(final String name) {
+        return new StorageDatabase(this.storage, new Key.From(name));
+    }
+
+    @Override
+    public Gauge gauge(final String name) {
+        return new StorageDatabase(this.storage, new Key.From(name));
+    }
+
+    /**
+     * Storage metrics database.
+     * @since 0.19
+     */
+    private static final class StorageDatabase implements Counter, Gauge {
+
+        /**
+         * Storage.
+         */
+        private final Storage storage;
+
+        /**
+         * Metrics key.
+         */
+        private final Key key;
+
+        /**
+         * New storage metrics database.
+         * @param storage Storage
+         * @param key Key
+         */
+        StorageDatabase(final Storage storage, final Key key) {
+            this.storage = storage;
+            this.key = key;
+        }
+
+        @Override
+        public void set(final long value) {
+            this.storage.save(this.key, content(value))
+                .handle(StorageDatabase::handle);
+        }
+
+        @Override
+        public void add(final long amount) {
+            this.storage.exists(this.key).thenCompose(
+                exists -> {
+                    final CompletionStage<Long> res;
+                    if (exists) {
+                        res = this.storage.value(this.key).thenCompose(
+                            content -> new PublisherAs(content)
+                                .string(StandardCharsets.UTF_8)
+                                .thenApply(Long::valueOf)
+                        );
+                    } else {
+                        res = CompletableFuture.completedFuture(0L);
+                    }
+                    return res;
+                }
+            ).thenCompose(val -> this.storage.save(this.key, content(val + amount)))
+                .handle(StorageDatabase::handle);
```

Comment:
@g4s8 this approach to updating values will lead to race conditions right away (the method is called at least once for each request, that's very often) and very high performance overhead (one memory update is 3 storage ops). I believe that buffering ans synchronizing writes is mandatory here from the start.

What do you think about collecting data to `InMemoryMetrics` and publishing it periodically to `Storage` approximately the same way we publish it to log? That way we have buffering and synchronized writes.